src/org/sosy_lab/java_smt/solvers/cvc5/CVC5InterpolatingProver.java

```diff
@@ -0,0 +1,370 @@
+   *
+   * @param formula List of Collections of formulas
+   * @param usingSolver the CVC5 Solver Instance to use
+   * @return concatenated Formulas with AND as CVC5 Term
+   */
+  private Term buildConjunctionOfFormulasOverList(
+      ImmutableList<Collection<Term>> formula, Solver usingSolver) {
+    Collection<Term> eachCollectionOfTermsConcat =
+        formula.stream()
+            .map(n -> buildConjunctionOfFormulas(n, usingSolver))
+            .collect(ImmutableList.toImmutableList());
+    Term concatTerm = buildConjunctionOfFormulas(eachCollectionOfTermsConcat, usingSolver);
+
+    return concatTerm;
+  }
+
+  /**
+   * Turns a collection of Formulas to a Single Conjunction of the Formulas e.g.: [A,B,C] ->
+   * A/\B/\C.
+   *
+   * @param formulas collection of formulas
+   * @param usingSolver the CVC5 Solver Instance to use
+   * @return concatenated Formulas with AND as CVC5 Term
+   */
+  private Term buildConjunctionOfFormulas(Collection<Term> formulas, Solver usingSolver) {
+    Preconditions.checkState(!closed);
+    Preconditions.checkArgument(!formulas.isEmpty());
+
+    Term formula = formulas.iterator().next();
+
+    Collection<Term> removedFormulas =
+        formulas.stream().filter(n -> !formula.equals(n)).collect(ImmutableList.toImmutableList());
+
+    if (removedFormulas.isEmpty()) {
+      return formula;
+    }
+
+    return usingSolver.mkTerm(
+        Kind.AND, formula, buildConjunctionOfFormulas(removedFormulas, usingSolver));
+  }
+
+  /**
+   * Generates Interpolation Pairs for Tree Interpolation. Experimental!
+   *
+   * @param pPartitionedFormulas of formulas
+   * @param pStartOfSubTree The start of the subtree containing the formula at this index as root.
+   * @return An List of Interpolation Pairs (as Tuple (a List)) containing Lists of Collection of
+   *     Terms
+   */
+  @SuppressWarnings("unchecked")
+  private ImmutableList<ImmutableList<ImmutableList<Collection<Term>>>> getTreeInterpolationPairs(
+      List<? extends Collection<Term>> pPartitionedFormulas, int[] pStartOfSubTree) {
+    // First Interpolation Pair
+    ImmutableList<Integer> leafes = ImmutableList.of(pStartOfSubTree[0]);
+    // current generated LHS of Tuple
+    ImmutableList<Collection<Term>> currA = ImmutableList.of(pPartitionedFormulas.get(0));
+    // current generated RHS of Tuple
+    ImmutableList<Collection<Term>> currB =
+        ImmutableList.copyOf(pPartitionedFormulas.subList(1, pPartitionedFormulas.size()));
+    // current generated Interpolation Tuple
+    ImmutableList<ImmutableList<Collection<Term>>> betweenRes = ImmutableList.of(currA, currB);
+    ImmutableList<ImmutableList<ImmutableList<Collection<Term>>>> result =
+        ImmutableList.of(betweenRes);
+    // iterate through Tree structure
+    for (int i = 1; i < pStartOfSubTree.length - 1; i++) {
+      // if the leave does not change, continue like Sequential Interpolation
+      if (pStartOfSubTree[i] == pStartOfSubTree[i - 1]) {
+        currA =
```

Comment:
Just because there are immutable data-structures and there are streams, this does not enforce you to only use that kind of data-structures. In this method, that kind of logic makes it really hard to understand and follow the program. It might be much easier to just use a plain Lists or Sets and add elements to them.

I still have no clue what this method is doing.

Reply:
ArrayLists, which were used in the beginning, led to Refaster complaining -> replace everything with ImmutableLists (I used the first found method for replacing ArrayLists operators with methods applicable for ImmutableList, mainly some kind of Stream, hence the unreadability). The order matters and i need to copy lists here and there, so ImmutableLists seemed like an alternative. 
This method tries to bring Tree Interpolation to Solvers, with no support for it. It returns a List of Interpolation Pairs (dependent on the given partitionedFormulas and startOfSubTree), following Tree Interpolation. In the following steps, a simple Interpolation Call is sufficient.

Reply:
Is each pair that is returned here defined as the "subtree" and the "rest of the whole tree", based on a single cut-point that is the interpolation position?

If this is the case, then the interpolation criteria might not be fulfilled.
The reasoning is quite similar to the comment lately regarding sequential interpolation with one proof vs several proofs. 

There is a example documented in CPAchecker, that was a problem for another tree-interpolation problem:
https://gitlab.com/sosy-lab/software/cpachecker/-/blob/trunk/src/org/sosy_lab/cpachecker/util/predicates/interpolation/strategy/WellScopedInterpolation.java
You might give it a try and produce tree interpolants for it.

Reply:
I dropped the idea of implementing `treeInterpolation` for CVC5. Maybe doing it properly for generally any solver would be wiser. Looking at the slides of yours below, it does not look to complicated, something I would be happy to try out.